### PR TITLE
envoy: Use github.com/cilium/proxy

### PR DESCRIPTION
**Pull request by jrajahalme**
Add a new variable `envoy.RequiredEnvoyVersionSHA` and check that the
running Envoy version matches that instead of the current Cilium SHA.

This requires that Envoy version dependency is manually updated, but
allows Cilium SHA to advance without requiring to rebuild Envoy.

Pull cilium-envoy from docker image `cilium/cilium-envoy:<CILIUM_ENVOY_SHA>`, where `<CILIUM_ENVOY_SHA>` is the git SHA of the dependency to the new github.com/cilium/proxy repo.

Use a new (experimental) builder image from which the Envoy build dependencies have been removed.

Pull in the generated protobuf API files via a vendor dependency to github.com/cilium/proxy. This is also referred to via a SHA, which may be different from the CILIUM_ENVOY_SHA due to the protobufs maybe changing less often.

Proxy dependency to the go build generated `libcilium.h` that captures the exported function prototypes from `proxylib/proxylib.go` is currently handled manually without any checks. This means that if `proxylib/proxylib.go` exported function prototypes are changed Envoy will fail when calling those functions.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6123)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Hit by https://github.com/cilium/cilium/issues/5745

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
`cilium-docker` seems to have built ok, but not found when installing: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/7739/execution/node/92/log/?consoleFull

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
CI: "Required context class hudson.FilePath is missing"

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
@eloycoto CI parallel builds are interfering with each other now that Envoy build has been removed. Will have to remove parallel `make clean` steps to fix that

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Added missing vendor files...

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Hit by #6122

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
@eloycoto Separated the last commit to a PR of its own: #6132 

Removing Envoy build from the CI made this problem more apparent, but we already had disabled proxylib's clean target because of this.

**Comment by jrajahalme** [comment]
Added a build time check to make sure CILIUM_ENVOY_SHA matches with `RequiredEnvoyVersionSHA` in `pkg/envoy/envoy.go`.

**Comment by jrajahalme** [comment]
test-me-please

File: Makefile [hunk 1]
```diff
@@ -124,23 +121,20 @@ install:
...
+	$(DOCKER) build --build-arg CILIUM_ENVOY_SHA=${CILIUM_ENVOY_SHA} --build-arg LOCKDEBUG=${LOCKDEBUG} --build-arg V=${V} -t "cilium/cilium:$(DOCKER_IMAGE_TAG)" .
 	$(QUIET)echo "Push like this when ready:"
 	$(QUIET)echo "docker push cilium/cilium:$(DOCKER_IMAGE_TAG)"
 
 docker-image-runtime:
 	cd contrib/packaging/docker && docker build -t "cilium/cilium-runtime:$(UTC_DATE)" -f Dockerfile.runtime .
 
 docker-image-builder:
-	cd envoy && docker build -t "quay.io/cilium/cilium-builder:$(UTC_DATE)" .
+	docker build -t "cilium/cilium-builder:jarno-test" -f Dockerfile.builder .
```
Comment (tgraf) [inline Makefile]: I assume the `jarno-test` is temporary and will be removed in the final submission?

File: pkg/envoy/envoy.go [hunk 2]
```diff
@@ -39,6 +39,7 @@ import (
 var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "envoy-manager")
 
 var (
+	RequiredEnvoyVersionSHA = "fd3866831d8f85e27120f34d98133343fc9c4e0e"
```
Comment (tgraf) [inline pkg/envoy/envoy.go:45]: Please add a comment that describes when to update this.

File: test/docker-compose.yml [hunk 3]
```diff
@@ -12,7 +12,7 @@ services:
     command: "etcd -name etcd0 -advertise-client-urls http://0.0.0.0:4002 -listen-client-urls http://0.0.0.0:4002 -initial-cluster-token etcd-cluster-1 -initial-cluster-state new"
     privileged: true
   base_image:
-    image: "quay.io/cilium/cilium-builder:2018-10-29"
+    image: "cilium/cilium-builder:jarno-test"
```
Comment (tgraf) [inline test/docker-compose.yml]: Temporary?

File: Gopkg.toml [hunk 4]
```diff
@@ -298,3 +298,7 @@ required = [
 [[constraint]]
   name = "github.com/hashicorp/go-immutable-radix"
   version = "1.0.0"
+
+[[constraint]]
+  name = "github.com/cilium/proxy"
```
Comment (tgraf) [inline Gopkg.toml:303]: Nice!

**Review by tgraf — CHANGES_REQUESTED** [review]
Awesome!

File: Makefile (hunk 1 above)
Comment (jrajahalme) [reply Makefile]: Right!

File: test/docker-compose.yml (hunk 3 above)
Comment (jrajahalme) [reply test/docker-compose.yml]: Yes, was testing with a stripped down builder. Can either produce official builder with the new Dockerfile, or use the old ones for now.

**Comment by jrajahalme** [comment]
Travis CI failed on the `proxylib/proxylib_test.go` due to the test accesslog server lagging behind, likely due to a resource constraint. Plan to change the tooling so that we wait longer if the expected number of messages has not been received, but not otherwise so that the tests do not get slower. Issued #6135

File: Dockerfile [hunk 5]
```diff
@@ -8,7 +17,7 @@
 # versions to be built while allowing the new versions to make changes
 # that are not backwards compatible.
 #
-FROM quay.io/cilium/cilium-builder:2018-10-29 as builder
+FROM cilium/cilium-builder:jarno-test as builder
```
Comment (aanm) [inline Dockerfile]: Change image tag

File: Dockerfile.builder [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#
```
Comment (aanm) [inline Dockerfile.builder:1]: Can you leave a description where is this image used?

File: Documentation/contributing.rst [hunk 7]
```diff
@@ -41,22 +41,10 @@ contribute to Cilium:
...
-+----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
 + `Docker <https://docs.docker.com/engine/installation/>`_                         | OS-Dependent             | N/A (OS-specific)                                                             |
 +----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
 + `Docker-Compose <https://docs.docker.com/compose/install/>`_                     | OS-Dependent             | N/A (OS-specific)                                                             |
 +----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
-+ `Cmake  <https://cmake.org/download/>`_                                          | OS-Dependent             | N/A (OS-specific)                                                             |
```
Comment (aanm) [inline Documentation/contributing.rst:52]: :tada:

File: test/run-envoy-unit-tests.sh [hunk 8]
```diff
@@ -1,5 +0,0 @@
-#!/bin/bash
-
-set -e
-
-vagrant ssh runtime -c "cd /home/vagrant/go/src/github.com/cilium/cilium ;make build; make tests-envoy"
```
Comment (aanm) [inline test/run-envoy-unit-tests.sh:5]: How will we test envoy from now on? Is there a plan to create a jenkins job for the new repo? cc @eloycoto

File: Makefile [hunk 9]
```diff
@@ -121,7 +121,9 @@ install:
 GIT_VERSION: .git
 	echo "$(GIT_VERSION)" >GIT_VERSION
 
-docker-image: clean GIT_VERSION
+docker-image: clean docker-image-no-clean
```
Comment (aanm) [inline Makefile:127]: This can be made in a different PR. I think @nebril wants this a lot.

**Review by aanm — CHANGES_REQUESTED** [review]
:tada: :tada: :tada:

File: test/run-envoy-unit-tests.sh (hunk 8 above)
Comment (jrajahalme) [reply test/run-envoy-unit-tests.sh:5]: No concrete plan yet, but obviously needs to be done.

File: Makefile (hunk 9 above)
Comment (jrajahalme) [reply Makefile:127]: Done & merged (#6132).

File: pkg/envoy/envoy.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/envoy/envoy.go:45]: Done.

File: Dockerfile (hunk 5 above)
Comment (jrajahalme) [reply Dockerfile]: Built the new builder image in quay.io, with tag 2018-11-01. It is now only 235MB ;-)

File: Dockerfile.builder (hunk 6 above)
Comment (jrajahalme) [reply Dockerfile.builder:1]: Added comment:
```
#
# Cilium build-time dependencies.
# Image created from this file is used to build Cilium.
#
```

Note that while I removed the most obvious Envoy dependencies from here, I'm not sure if all that remains is really needed to build Cilium. Maybe you can have look?

Comment (jrajahalme) [reply Dockerfile.builder:1]: @aanm ^^^^

File: test/run-envoy-unit-tests.sh (hunk 8 above)
Comment (jrajahalme) [reply test/run-envoy-unit-tests.sh:5]: So far there are no code changes in the cilium/proxy repo, so we are good.

**Comment by jrajahalme** [comment]
test-me-please

File: Dockerfile.builder (hunk 6 above)
Comment (aanm) [reply Dockerfile.builder:1]: I know it's used to build Cilium :-) I'm just not sure if this one is the cilium/builder or the cilium/runtime

**Review by ianvernon — APPROVED** [review]

**Review by tgraf — APPROVED** [review]
